spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractExecutorChannel.java

```diff
@@ -0,0 +1,221 @@
+ */
+
+package org.springframework.integration.channel;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executor;
+
+import org.springframework.integration.dispatcher.AbstractDispatcher;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageDeliveryException;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessagingException;
+import org.springframework.messaging.support.ChannelInterceptor;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+import org.springframework.messaging.support.MessageHandlingRunnable;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link AbstractSubscribableChannel} base implementation for those inheritors
+ * which logic may be based on the {@link Executor}.
+ * <p>
+ * Utilizes common operations for the {@link AbstractDispatcher}.
+ * <p>
+ * Implements the {@link ExecutorChannelInterceptor}s logic when the message handling
+ * is handed to the {@link Executor#execute(Runnable)}.
+ *
+ * @author Artem Bilan
+ * @see ExecutorChannel
+ * @see PublishSubscribeChannel
+ * @since 4.2
+ */
+public abstract class AbstractExecutorChannel extends AbstractSubscribableChannel {
+
+	protected final List<ExecutorChannelInterceptor> executorInterceptors =
+			new CopyOnWriteArrayList<ExecutorChannelInterceptor>();
+
+	protected volatile Executor executor;
+
+	protected volatile AbstractDispatcher dispatcher;
+
+	protected volatile Integer maxSubscribers;
+
+	protected volatile int executorInterceptorsSize;
+
+	public AbstractExecutorChannel(Executor executor) {
+		this.executor = executor;
+	}
+
+	/**
+	 * Specify the maximum number of subscribers supported by the
+	 * channel's dispatcher.
+	 *
+	 * @param maxSubscribers The maximum number of subscribers allowed.
+	 */
+	public void setMaxSubscribers(int maxSubscribers) {
+		this.maxSubscribers = maxSubscribers;
+		this.dispatcher.setMaxSubscribers(maxSubscribers);
+	}
+
+	@Override
+	public void setInterceptors(List<ChannelInterceptor> interceptors) {
+		super.setInterceptors(interceptors);
+		this.executorInterceptors.clear();
+		for (ChannelInterceptor interceptor : interceptors) {
+			if (interceptor instanceof ExecutorChannelInterceptor) {
+				this.executorInterceptors.add((ExecutorChannelInterceptor) interceptor);
+			}
+		}
+		this.executorInterceptorsSize = this.executorInterceptors.size();
+	}
+
+	@Override
+	public void addInterceptor(ChannelInterceptor interceptor) {
+		super.addInterceptor(interceptor);
+		if (interceptor instanceof ExecutorChannelInterceptor) {
+			this.executorInterceptors.add((ExecutorChannelInterceptor) interceptor);
+			this.executorInterceptorsSize++;
+		}
+	}
+
+	@Override
+	public void addInterceptor(int index, ChannelInterceptor interceptor) {
+		super.addInterceptor(index, interceptor);
```

Comment:
This is a little peculiar - does the index represent the position in super interceptors or executor interceptors?

Maybe we should only support 0 as the index here?